Editing entities from select2.

miina opened the pull request:
Fixes #8

miina committed:
Start editing from select2

miina committed:
Continue with select2 edit.

miina committed:
Fix scrolling not working when returning.

westonruter commented on js/customize-object-selector-component.js:
```diff
@@ -85,9 +85,11 @@ wp.customize.ObjectSelectorComponent = (function( api, $ ) {
 						return $.trim( component.select2_result_template( data ) );
 					},
 					templateSelection: function( data ) {
+						data.disable_edit = ! data.id;
```
The `id` is not defined when the template is for the “searching” message or for the “no results found” message?

miina replied:
Yes, and for placeholders as well.

westonruter commented on php/class-plugin.php:
```diff
@@ -391,8 +391,12 @@ public function print_templates() {
 					{{{ data.text }}}
 				</span>
-			<# } else { #>
+			<# } else if ( data.disable_edit ) { #>
```
Would it be better to just make this `else if ( ! data.id )`, or identify why specifically there is no `id` present? We could then eliminate the `disable_edit` property.

miina replied:
In case of search result `data.id` is present for the found post but the editing should be disabled in the selection list. I'll check the other existing properties and see if there's something to use instead of `disable_edit`.

miina committed:
Use data.element instead of data.edit_disabled

miina commented:
@westonruter Found that `data.element` is present only in case the object has to be link. Removed the `disable_edit` and replaced with using `element`.

valendesigns commented:
@miina This is working exactly how I had envisioned it, but I found an obscure bug. If you drag & drop the order of the posts, then edit one and return focus back to the control the Customizer panel will not scroll up or down. I cannot reproduce the issue without reordering the posts first.

westonruter commented:
Instead of making the result text itself a link, what if there was a link added _after_ the text that had the [edit dashicon](https://developer.wordpress.org/resource/dashicons/#edit) as its content (along with `screen-reader-text`)? This would ensure that the main text itself could still have the `cursor:move` whereas the link itself to edit could have `cursor:pointer`.

miina committed:
Add dashicon. Remove event delegation.

miina committed:
Remove debugging info.

valendesigns committed:
Clean up styles

valendesigns committed:
Fix phpcs issue

miina committed:
Fix scrolling issue.

miina commented:
I think the scrolling issue is fixed.

westonruter commented:
FYI: This same functionality is also proposed for the `dropdown-pages` core control: https://github.com/xwp/wp-customize-posts/issues/271

westonruter commented on js/customize-object-selector-component.js:
```diff
@@ -366,6 +366,47 @@ wp.customize.ObjectSelectorComponent = (function( api, $ ) {
+
+			// Set up the add new post buttons
+			component.container.on( 'click', '.select2-selection__choice__edit', function() {
+				var ensuredPromise, returnPromise, postId, $elm = $( this );
```
The body of this function can now be replaced with a call to `api.Posts.startEditPostFlow()`. See https://github.com/xwp/wp-customize-posts/blob/33840418da9dcd7c55d15061de5fe49f85d9058f/js/customize-posts.js#L589-L598

miina replied:
Thanks for pointing that out, will take a look.

westonruter commented:
As part of this I suggest the style of the create and edit buttons get updated to match the `dropdown-pages` now in Customize Posts:

<img width="299" alt="screen shot 2016-09-17 at 3 05 00 pm" src="https://cloud.githubusercontent.com/assets/134745/18728957/89feefb6-8004-11e6-8421-80e85eec7d60.png">

It's using dashicons for the icons and using flexbox to lay-out the controls.

miina committed:
Merge develop. Use startEditPostFlow.

miina committed:
Fix cs.

miina committed:
Match adding new button with Customize Posts.

miina committed:
Change edit button according to Customize Posts.

miina commented on js/customize-object-selector-component.js:
```diff
@@ -111,6 +111,7 @@ wp.customize.ObjectSelectorComponent = (function( api, $ ) {
 						return $.trim( component.select2_result_template( data ) );
 					},
 					templateSelection: function( data ) {
+						data.multiple = component.select2_options.multiple;
```
Is there a better way how to get it from template that it's multiple select field?

westonruter replied:
It doesn't look like. I checked `this` and if there are any other arguments, and I see none. So yeah, looks good.

miina committed:
Remove unused focusConstructWithBreadcrumb.

valendesigns committed:
Fix edit button visibility & position, and add new button text

valendesigns committed:
Fix eslint issues
